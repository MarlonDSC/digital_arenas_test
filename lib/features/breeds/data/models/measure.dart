import 'package:json_annotation/json_annotation.dart';

part 'measure.g.dart';

@JsonSerializable()
class Measure {
  final String imperial;
  final String metric;

  Measure({required this.imperial, required this.metric});

  factory Measure.fromJson(Map<String, dynamic> json) => _$MeasureFromJson(json);
  Map<String, dynamic> toJson() => _$MeasureToJson(this);
}

