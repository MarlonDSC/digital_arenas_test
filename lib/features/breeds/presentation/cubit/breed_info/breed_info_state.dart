part of 'breed_info_cubit.dart';

sealed class BreedInfoState extends Equatable {
  const BreedInfoState();

  @override
  List<Object> get props => [];
}

final class BreedInfoInitial extends BreedInfoState {}

final class BreedInfoLoading extends BreedInfoState {}

final class BreedInfoLoaded extends BreedInfoState {
  final BreedInfo breedInfo;

  const BreedInfoLoaded({required this.breedInfo});

  @override
  List<Object> get props => [breedInfo];
}

final class BreedInfoError extends BreedInfoState {
  final List<BaseError> errors;

  const BreedInfoError({required this.errors});

  @override
  List<Object> get props => [errors];
}
