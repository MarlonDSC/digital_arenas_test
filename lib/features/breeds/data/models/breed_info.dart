import 'package:inmo_mobile/features/breeds/data/models/measure.dart';
import 'package:inmo_mobile/features/favourite/data/models/favourite.dart';
import 'package:json_annotation/json_annotation.dart';

part 'breed_info.g.dart';

@JsonSerializable()
class BreedInfo {
  final int id;
  final String name;
  final Measure weight;
  final Measure height;
  @JsonKey(name: 'life_span')
  final String? lifeSpan;
  @JsonKey(name: 'bred_for')
  final String? bredFor;
  @JsonKey(name: 'breed_group')
  final String? breedGroup;
  final String? temperament;
  @JsonKey(name: 'reference_image_id')
  final String referenceImageId;

  BreedInfo({
    required this.id,
    required this.name,
    required this.weight,
    required this.height,
    required this.lifeSpan,
    required this.bredFor,
    required this.breedGroup,
    required this.temperament,
    required this.referenceImageId,
  });

  factory BreedInfo.fromFavourite(Favourite favourite) {
    return BreedInfo(
      id: favourite.breedId,
      name: favourite.breedName,
      weight: Measure(
        imperial: favourite.breedImperialWeight,
        metric: favourite.breedMetricWeight,
      ),
      height: Measure(
        imperial: favourite.breedImperialHeight,
        metric: favourite.breedMetricHeight,
      ),
      lifeSpan: favourite.breedLifeSpan,
      bredFor: favourite.breedBredFor,
      breedGroup: favourite.breedBreedGroup,
      temperament: favourite.breedTemperament,
      referenceImageId: favourite.breedImageId,
    );
  }

  factory BreedInfo.fromJson(Map<String, dynamic> json) =>
      _$BreedInfoFromJson(json);
  Map<String, dynamic> toJson() => _$BreedInfoToJson(this);
}
