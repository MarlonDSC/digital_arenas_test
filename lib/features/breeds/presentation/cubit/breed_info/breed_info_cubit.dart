import 'package:bloc/bloc.dart';
import 'package:equatable/equatable.dart';
import 'package:inmo_mobile/core/errors/base_error.dart';
import 'package:inmo_mobile/features/breeds/data/models/breed_info.dart';
import 'package:inmo_mobile/features/breeds/domain/usecases/get_breed_info.dart';

part 'breed_info_state.dart';

class BreedInfoCubit extends Cubit<BreedInfoState> {
  final GetBreedInfoUseCase _getBreedInfoUseCase;

  BreedInfoCubit(this._getBreedInfoUseCase) : super(BreedInfoInitial());

  Future<void> getBreedInfo(int id) async {
    emit(BreedInfoLoading());
    final result = await _getBreedInfoUseCase(id);
    result.fold(
      (l) => emit(BreedInfoError(errors: l)),
      (r) => emit(BreedInfoLoaded(breedInfo: r)),
    );
  }
}
