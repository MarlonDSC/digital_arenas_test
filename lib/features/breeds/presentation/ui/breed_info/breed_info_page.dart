import 'package:flutter/material.dart';
import 'package:flutter_bloc/flutter_bloc.dart';
import 'package:inmo_mobile/core/errors/bloc/modal/modal_bloc.dart';
import 'package:inmo_mobile/di.dart';
import 'package:inmo_mobile/features/breeds/data/models/breed_info.dart';
import 'package:inmo_mobile/features/breeds/presentation/cubit/breed_image/breed_image_cubit.dart';
import 'package:inmo_mobile/features/breeds/presentation/cubit/breed_info/breed_info_cubit.dart';
import 'package:inmo_mobile/features/breeds/presentation/ui/breed_info/widgets/breed_card.dart';
import 'package:inmo_mobile/features/breeds/presentation/ui/breed_info/widgets/favourite_breed_sheet.dart';
import 'package:inmo_mobile/features/favourite/presentation/cubit/favourite_info/favourite_info_cubit.dart';

class BreedInfoPage extends StatefulWidget {
  final int id;
  const BreedInfoPage({super.key, required this.id});

  @override
  State<BreedInfoPage> createState() => _BreedInfoPageState();
}

class _BreedInfoPageState extends State<BreedInfoPage> {
  @override
  Widget build(BuildContext context) {
    return BlocConsumer<BreedInfoCubit, BreedInfoState>(
      listener: _onGetBreedInfo,
      builder: (context, state) {
        return BlocBuilder<BreedInfoCubit, BreedInfoState>(
          builder: (context, state) {
            if (state is BreedInfoLoading) {
              return const Center(child: CircularProgressIndicator());
            }
            if (state is BreedInfoLoaded) {
              return Column(
                children: [
                  BreedCard(breedInfo: state.breedInfo),
                  ElevatedButton.icon(
                    onPressed: () {
                      _showBottomSheet(context, state.breedInfo);
                    },
                    icon: const Icon(Icons.favorite),
                    label: const Text('AÃ±adir a favoritos'),
                  ),
                ],
              );
            } else if (state is BreedInfoError) {
              return ElevatedButton(
                onPressed: () {
                  context.read<BreedInfoCubit>().getBreedInfo(widget.id);
                },
                child: const Text('Try again'),
              );
            }
            return const SizedBox.shrink();
          },
        );
      },
    );
  }

  void _onGetBreedInfo(BuildContext context, BreedInfoState state) {
    if (state is BreedInfoError) {
      ScaffoldMessenger.of(
        context,
      ).showSnackBar(SnackBar(content: Text(state.errors.toString())));
    } else if (state is BreedInfoLoaded) {
      context.read<BreedImageCubit>().getBreedImage(
        state.breedInfo.referenceImageId,
      );
    }
  }

  void _showBottomSheet(BuildContext context, BreedInfo breedInfo) {
    context.read<ModalBloc>().add(
      DisplayModal(
        BlocProvider(
          create: (context) => sl<FavouriteInfoCubit>(),
          child: FavouriteBreedSheet(
            breedInfo: breedInfo,
          ),
        ),
      ),
    );
  }
}
