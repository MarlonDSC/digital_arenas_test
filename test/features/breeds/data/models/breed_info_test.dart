import 'package:flutter_test/flutter_test.dart';
import 'package:approval_tests/approval_tests.dart';
import 'package:inmo_mobile/features/breeds/data/models/breed_info.dart';
import 'package:inmo_mobile/features/breeds/data/models/measure.dart';

void main() {
  group('BreedInfo', () {
    test('verify basic serialization', () {
      final breedInfo = BreedInfo(
        id: 1,
        name: 'Golden Retriever',
        weight: Measure(imperial: '55-75 lbs', metric: '25-34 kg'),
        height: Measure(imperial: '21.5-24 inches', metric: '55-61 cm'),
        lifeSpan: '10-12 years',
        bredFor: 'Retrieving game for hunters',
        breedGroup: 'Sporting',
        temperament: 'Intelligent, Kind, Reliable',
        referenceImageId: 'B1iWlZRXQ',
      );

      Approvals.verifyAsJson(
        breedInfo,
        options: const Options(
          reporter: DiffReporter(),
        ),
      );
    });

    test('verify serialization with null optional fields', () {
      final breedInfo = BreedInfo(
        id: 2,
        name: 'Mixed Breed',
        weight: Measure(imperial: '30-50 lbs', metric: '14-23 kg'),
        height: Measure(imperial: '18-22 inches', metric: '46-56 cm'),
        lifeSpan: null,
        bredFor: null,
        breedGroup: null,
        temperament: null,
        referenceImageId: 'B2iWlZRXQ',
      );

      Approvals.verifyAsJson(
        breedInfo,
        options: const Options(
          reporter: DiffReporter(),
        ),
      );
    });

    test('verify fromJson factory with sample payload', () {
      const json = {
        'id': 4,
        'name': 'German Shepherd',
        'weight': {
          'imperial': '50-90 lbs',
          'metric': '23-41 kg',
        },
        'height': {
          'imperial': '22-26 inches',
          'metric': '56-66 cm',
        },
        'life_span': '7-10 years',
        'bred_for': 'Herding sheep',
        'breed_group': 'Herding',
        'temperament': 'Loyal, Confident, Courageous',
        'reference_image_id': 'B4iWlZRXQ',
      };

      final breedInfo = BreedInfo.fromJson(json);

      Approvals.verifyAsJson(
        {
          'input': json,
          'parsed': breedInfo,
        },
        options: const Options(
          reporter: DiffReporter(),
        ),
      );
    });
  });
}