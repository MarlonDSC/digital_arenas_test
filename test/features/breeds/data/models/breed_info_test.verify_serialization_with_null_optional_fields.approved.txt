# This file was generated by approval_tests. Please do not edit.

{
  "breedInfo": {
    "id": 2,
    "name": "Mixed Breed",
    "weight": {
      "imperial": "30-50 lbs",
      "metric": "14-23 kg"
    },
    "height": {
      "imperial": "18-22 inches",
      "metric": "46-56 cm"
    },
    "life_span": null,
    "bred_for": null,
    "breed_group": null,
    "temperament": null,
    "reference_image_id": "B2iWlZRXQ"
  }
}